File: backend/core/transaction_types.py
from typing import Any

from core.http import api_error


class TransactionTypes:
    GENERIC = "GENERIC"
    P2P = "P2P_TRANSFER"
    BOND_PURCHASE = "BOND_PURCHASE"
    BOND_TRANSFER = "BOND_TRANSFER"
    BOND_REDEMPTION = "BOND_REDEMPTION"
    BOND_MATURITY_SETTLEMENT = "BOND_MATURITY_SETTLEMENT"
    MOBILE_RECHARGE = "MOBILE_RECHARGE"
    DTH_RECHARGE = "DTH_RECHARGE"
    FASTAG_RECHARGE = "FASTAG_RECHARGE"
    ELECTRICITY_BILL = "ELECTRICITY_BILL_PAYMENT"
    UTILITY_PAYMENT = "UTILITY_PAYMENT"


TARGET_TYPE_RULES: dict[str, set[str]] = {
    TransactionTypes.MOBILE_RECHARGE: {"operator", "mobileNumber", "type"},
    TransactionTypes.DTH_RECHARGE: {"provider", "dthNumber"},
    TransactionTypes.FASTAG_RECHARGE: {"bank", "vehicleNumber"},
    TransactionTypes.ELECTRICITY_BILL: {"state", "board", "consumerNumber"},
}


def resolve_intent_transaction_type(target_type: str) -> str:
    normalized = (target_type or "").strip().upper()
    if not normalized:
        api_error(400, "INVALID_TARGET_TYPE", "target_type is required.")
    if normalized in TARGET_TYPE_RULES:
        return normalized
    # Allow unknown types but classify under UTILITY_PAYMENT.
    return TransactionTypes.UTILITY_PAYMENT


def validate_intent_metadata(target_type: str, metadata: dict[str, Any] | None):
    normalized = (target_type or "").strip().upper()
    required = TARGET_TYPE_RULES.get(normalized)
    if not required:
        return
    if not isinstance(metadata, dict):
        api_error(400, "INVALID_METADATA", f"transaction_metadata is required for {normalized}.")
    missing = [field for field in sorted(required) if field not in metadata or metadata[field] in (None, "")]
    if missing:
        api_error(400, "MISSING_METADATA_FIELDS", "Missing required transaction metadata fields.", missing_fields=missing)

File: frontend/lib/transactionTypes.ts
export type TransactionRecord = {
  id: string;
  description: string;
  transaction_type?: string;
  direction: "DEBIT" | "CREDIT";
  amount_inr: number;
  status: string;
  date: string;
  metadata?: Record<string, unknown>;
};

const TYPE_LABELS: Record<string, string> = {
  P2P_TRANSFER: "P2P Transfer",
  MOBILE_RECHARGE: "Mobile Recharge",
  DTH_RECHARGE: "DTH Recharge",
  FASTAG_RECHARGE: "FASTag Recharge",
  ELECTRICITY_BILL_PAYMENT: "Electricity Bill",
  UTILITY_PAYMENT: "Utility Payment",
  BOND_PURCHASE: "Bond Purchase",
  BOND_TRANSFER: "Bond Transfer",
  BOND_REDEMPTION: "Bond Redemption",
  BOND_MATURITY_SETTLEMENT: "Maturity Settlement",
  GENERIC: "Generic",
};

const METADATA_LABELS: Record<string, string> = {
  operator: "Operator",
  mobileNumber: "Mobile Number",
  type: "Plan Type",
  provider: "Provider",
  dthNumber: "Subscriber ID",
  bank: "Issuing Bank",
  vehicleNumber: "Vehicle Number",
  state: "State",
  board: "Electricity Board",
  consumerNumber: "Consumer Number",
  recipient_mobile: "Recipient Mobile",
  amount_paise: "Amount (Paise)",
  target_type: "Target Type",
};

const METADATA_ORDER: Record<string, string[]> = {
  MOBILE_RECHARGE: ["operator", "mobileNumber", "type"],
  DTH_RECHARGE: ["provider", "dthNumber"],
  FASTAG_RECHARGE: ["bank", "vehicleNumber"],
  ELECTRICITY_BILL_PAYMENT: ["state", "board", "consumerNumber"],
  P2P_TRANSFER: ["recipient_mobile", "amount_paise"],
};

export function formatTransactionType(type?: string): string {
  if (!type) return "Generic";
  return TYPE_LABELS[type] || type.replace(/_/g, " ");
}

export function getMetadataRows(
  targetType: string | undefined,
  metadata: Record<string, unknown> | undefined
): Array<{ key: string; label: string; value: string }> {
  if (!metadata) return [];
  const keys = Object.keys(metadata);
  if (keys.length === 0) return [];

  const orderedKeys = METADATA_ORDER[targetType || ""] || keys;
  const remaining = keys.filter((k) => !orderedKeys.includes(k));
  const finalKeys = [...orderedKeys.filter((k) => keys.includes(k)), ...remaining];

  return finalKeys.map((key) => ({
    key,
    label: METADATA_LABELS[key] || key.replace(/([A-Z])/g, " $1").trim(),
    value: String(metadata[key]),
  }));
}

